shared_model/interfaces/iroha_internal/abstract_transport_factory.hpp

```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_ABSTRACT_TRANSPORT_FACTORY_HPP
+#define IROHA_ABSTRACT_TRANSPORT_FACTORY_HPP
+
+#include <memory>
+
+#include "common/result.hpp"
+#include "cryptography/hash.hpp"
+#include "interfaces/common_objects/types.hpp"
+
+namespace shared_model {
+  namespace interface {
+
+    template <typename Interface, typename Transport>
+    class AbstractTransportFactory {
+     public:
+      struct Error {
+        types::HashType hash;
+        std::string error;
+      };
+
+      virtual iroha::expected::Result<std::unique_ptr<Interface>, Error> build(
+          Transport transport) const = 0;
```

Comment:
Why not to take const ref of transport?

Reply:
That will require to copy the argument regardless of the context. Value allows moving the argument if required.